src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,589 @@
+         * @implSpec If this method returns {@code false} then no further
+         * elements will be accepted and subsequent invocations of this method
+         * will return {@code false}.
+         *
+         * @param element the element to push downstream
+         * @return {@code true} if more elements can be sent,
+         *         and {@code false} if not.
+         */
+        boolean push(T element);
+
+        /**
+         * Allows for checking whether the next stage is known to not want
+         * any more elements sent to it.
+         *
+         * @apiNote This is best-effort only, once this returns {@code true} it
+         * should never return {@code false} again for the same instance.
+         *
+         * @implSpec By default this method returns {@code false}.
+         *
+         * @return {@code true} if this Downstream is known not to want any
+         *         more elements sent to it, {@code false} if otherwise
+         */
+        default boolean isRejecting() { return false; }
+    }
+
+    /**
+     * An Integrator receives elements and processes them,
+     * optionally using the supplied state, and optionally sends incremental
+     * results downstream.
+     *
+     * @param <A> the type of state used by this integrator
+     * @param <T> the type of elements this integrator consumes
+     * @param <R> the type of results this integrator can produce
+     * @since 22
+     */
+    @FunctionalInterface
+    @PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+    interface Integrator<A, T, R> {
+        /**
+         * Performs an action given: the current state, the next element, and
+         * a downstream object; potentially inspecting and/or updating
+         * the state, optionally sending any number of elements downstream
+         * -- and then returns whether more elements are to be consumed or not.
+         *
+         * @param state The state to integrate into
+         * @param element The element to integrate
+         * @param downstream The downstream object of this integration
+         * @return {@code true} if subsequent integration is desired,
+         *         {@code false} if not
+         */
+        boolean integrate(A state, T element, Downstream<? super R> downstream);
+
+        /**
+         * Factory method for turning Integrator-shaped lambdas into
+         * Integrators.
+         *
+         * @param integrator a lambda to create as Integrator
+         * @return the given lambda as an Integrator
+         * @param <A> the type of elements this integrator receives
+         * @param <T> the type of initializer this integrator consumes
+         * @param <R> the type of results this integrator can produce
+         */
+        @ForceInline
+        static <A, T, R> Integrator<A, T, R> of(Integrator<A, T, R> integrator) {
+            return integrator;
+        }
+
+        /**
+         * Factory method for turning Integrator-shaped lambdas into
```

Comment:
```suggestion
         * Targets Integrator-shaped lambdas to
```
It's not a factory, as it does produce anything, it's an identity function that aids in the targeting of the lambda expression to the integrator when there would otherwise be ambiguity as to the target.